### PR TITLE
[Fix] Prevent fields name conflicts with composite subfields names

### PR DESCRIPTION
**Pull request by ijreilly**
At field creation we are checking the availability of the name by comparing it to the other fields' names' on the object; but for composite fields the fields' names' as indicated in the repository do not exactly match the column names' on the tables (e.g "createdBy" field is actually represented by columns createdByName, createdBySource etc.).

### CODE REVIEW

**Review by charlesBochet — COMMENTED** [review]
I'm not sure it works!
If I create a custom field of type Link: 'example', this will use examplePrimaryLinkLabel column. Then I cannot create a 'examplePrimaryLinkLabel' custom field of type Text on this object.

I feel we need to generate the list of used columns dynamically from the object metadata to handle this case (and also it would be easier to maintain for standard field too as it's automatically generated). Maybe a generateInavailableFieldName(objectMetadata) ?

File: packages/twenty-server/src/engine/metadata-modules/utils/validate-name-availability.utils.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
+import { computeCompositeColumnName } from 'src/engine/metadata-modules/field-metadata/utils/compute-column-name.util';
+import { isCompositeFieldMetadataType } from 'src/engine/metadata-modules/field-metadata/utils/is-composite-field-metadata-type.util';
+import { ObjectMetadataEntity } from 'src/engine/metadata-modules/object-metadata/object-metadata.entity';
+
+const getReservedCompositeFieldsNames = (
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/metadata-modules/utils/validate-name-availability.utils.ts:6]: getReservedCompositeFieldNames

File: packages/twenty-server/src/engine/metadata-modules/utils/validate-name-availability.utils.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+) => {
+  const reservedCompositeFieldsNames =
+    getReservedCompositeFieldsNames(objectMetadata);
+
+  if (reservedCompositeFieldsNames.includes(name)) {
+    throw new NameNotAvailableException(name);
+  }
+};
+
+export class NameNotAvailableException extends Error {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/metadata-modules/utils/validate-name-availability.utils.ts:39]: I think this should not be in the utils but in an Exception folder (see other exception examples)

File: packages/twenty-server/src/engine/metadata-modules/utils/validate-name-availability.utils.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
+import { computeCompositeColumnName } from 'src/engine/metadata-modules/field-metadata/utils/compute-column-name.util';
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/metadata-modules/utils/validate-name-availability.utils.ts:2]: great that you are re-using it!

**Review by charlesBochet — APPROVED** [review]
LGTM!

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/field-metadata.service.ts [hunk 4]
```diff
@@ -673,10 +684,14 @@ export class FieldMetadataService extends TypeOrmQueryService<FieldMetadataEntit
 
   private validateFieldMetadataInput<
     T extends UpdateFieldInput | CreateFieldInput,
-  >(fieldMetadataInput: T): T {
+  >(fieldMetadataInput: T, objectMetadata: ObjectMetadataEntity): T {
     if (fieldMetadataInput.name) {
       try {
         validateMetadataNameOrThrow(fieldMetadataInput.name);
+        validateNameAvailabilityOrThrow(
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/field-metadata.service.ts:691]: not sure to completely understand this validateMetadataNameOrThrow vs validateNameAvailabilityOrThrow from the outside
I feel they should be wrapped together
maybe validateMetadataNameOrThrow should call validateNameAvailabilityOrThrow internally?

Comment (ijreilly) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/field-metadata.service.ts]: I updated the functions names to add clarity, but one cannot call another as validateMetadataNameOrThrow (renamed to validateMetadataNameValidityOrThrow) checks for the correctness of the string (only contains non-latin char, does not start with digit etc) and is used in other services to check all metadata names (for objects and relations too)

**Pull request merged**